File: packages/compiler/src/transforms/vShow.ts
import { IRNodeTypes } from '../ir'
import { createCompilerError, ErrorCodes, resolveDirective } from '../utils'
import type { DirectiveTransform } from '../transform'

export const transformVShow: DirectiveTransform = (_dir, node, context) => {
  const dir = resolveDirective(_dir, context)
  const { exp, loc } = dir
  if (!exp) {
    context.options.onError(
      createCompilerError(ErrorCodes.X_V_SHOW_NO_EXPRESSION, loc as any),
    )
    return
  }

  context.registerOperation({
    type: IRNodeTypes.DIRECTIVE,
    element: context.reference(),
    dir,
    name: 'show',
    builtin: true,
  })
}

File: packages/compiler-rs/src/transform/v_show.rs
use napi::bindgen_prelude::Either16;
use oxc_ast::ast::{JSXAttribute, JSXElement};

use crate::{
  ir::index::{BlockIRNode, DirectiveIRNode, SimpleExpressionNode},
  transform::{DirectiveTransformResult, TransformContext},
  utils::{directive::resolve_directive, error::ErrorCodes},
};

pub fn transform_v_show<'a>(
  _dir: &'a mut JSXAttribute<'a>,
  _: &JSXElement,
  context: &'a TransformContext<'a>,
  context_block: &'a mut BlockIRNode<'a>,
) -> Option<DirectiveTransformResult<'a>> {
  let mut dir = resolve_directive(_dir, context);
  if dir.exp.is_none() {
    context.options.on_error.as_ref()(ErrorCodes::VShowNoExpression);
    dir.exp = Some(SimpleExpressionNode::default())
  }

  let element = context.reference(&mut context_block.dynamic);
  context.register_operation(
    context_block,
    Either16::M(DirectiveIRNode {
      directive: true,
      element,
      dir,
      name: String::from("show"),
      builtin: Some(true),
      asset: None,
      model_type: None,
    }),
    None,
  );
  None
}
